### PR TITLE
Handle symbol name conflicts during executable linking.

### PR DESCRIPTION
**Pull request by ScottTodd**
Executable linking merges multiple modules together into a single module. If a symbol name already exists in the merged module, we previously skipped over it assuming the contents of those symbols were identical. This now resolve conflicts by
* renaming different private symbols
* merging equivalent private symbols
* emitting an error for public symbol conflicts

Fixes https://github.com/google/iree/issues/5946.
Includes a step towards https://github.com/google/iree/issues/4670 that should be made generic in some way.
Issues like https://github.com/google/iree/issues/4897 will still need to use MLIR symbol visibility (not some special visibility) or ensure uniqueness prior to this pass.

### CODE REVIEW

File: iree/compiler/Dialect/HAL/Target/TargetBackend.cpp [hunk 1]
```diff
@@ -102,33 +102,85 @@ void TargetBackend::declareTargetOps(IREE::Flow::ExecutableOp sourceOp,
...
+          }
+        } else {
+          // The source symbol has 'nested' or 'public' visibility.
+          if (SymbolTable::getSymbolVisibility(targetOp) !=
+              SymbolTable::Visibility::Private) {
+            // Oops! Both symbols are public and we can't safely rename either.
+            // If you hit this with ops that you think are safe to rename, mark
+            // them private.
+            return op->emitError()
+                   << "multiple public symbols with the name: " << symbolName;
```
Comment (ScottTodd) [inline iree/compiler/Dialect/HAL/Target/TargetBackend.cpp:156]: We could also skip linking between executables with conflicting public symbol names. I guess it depends on what sorts of cases where there could be conflicts and if an error would be more appropriate.

File: iree/compiler/Dialect/HAL/Target/TargetBackend.cpp [hunk 2]
```diff
@@ -102,33 +102,85 @@ void TargetBackend::declareTargetOps(IREE::Flow::ExecutableOp sourceOp,
...
-    if (auto symbolInterface = dyn_cast<SymbolOpInterface>(op)) {
-      if (targetSymbolMap.count(symbolInterface.getName())) {
-        // TODO(scotttodd): compare ops to ensure we aren't copying different
-        // things with the same name.
-        continue;
+    if (auto symbolOp = dyn_cast<SymbolOpInterface>(op)) {
+      auto symbolName = symbolOp.getName();
+
+      // Resolve symbol name conflicts.
+      if (targetSymbolMap.count(symbolName)) {
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/HAL/Target/TargetBackend.cpp:128]: You could do 

```
auto targetOp = targetSymbolMap.lookup(symbolName)
if (targetOp) {

}
```

File: iree/compiler/Dialect/HAL/Target/TargetBackend.cpp (hunk 1 above)
Comment (MaheshRavishankar) [reply iree/compiler/Dialect/HAL/Target/TargetBackend.cpp:156]: If this doesnt fail today then reporting error is fine. It is a conflicting symbol, so error seems like a good option.

File: iree/compiler/Dialect/HAL/Target/TargetBackend.cpp [hunk 3]
```diff
@@ -102,33 +102,85 @@ void TargetBackend::declareTargetOps(IREE::Flow::ExecutableOp sourceOp,
...
+          if (SymbolTable::getSymbolVisibility(targetOp) !=
+              SymbolTable::Visibility::Private) {
+            // Oops! Both symbols are public and we can't safely rename either.
+            // If you hit this with ops that you think are safe to rename, mark
+            // them private.
+            return op->emitError()
+                   << "multiple public symbols with the name: " << symbolName;
+          } else {
+            // Keep the original name for our new op, rename the target op.
+            auto disambiguateName =
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/HAL/Target/TargetBackend.cpp:159]: Move this into a function and call it from here? Its used in two place here.

File: iree/compiler/Dialect/VM/IR/VMOps.cpp [hunk 4]
```diff
@@ -744,11 +754,19 @@ static ParseResult parseRodataOp(OpAsmParser &parser, OperationState *result) {
...
     return failure();
   }
+
   return success();
 }
 
 static void printRodataOp(OpAsmPrinter &p, RodataOp &op) {
   p << op.getOperationName() << ' ';
+
+  // TODO(#4670): Share across ops or upstream a custom directive
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/VM/IR/VMOps.cpp:764]: I see we dont print all the attributes on this op. The canonical way of doing this is to print the attribute list on the op, there is a stock method to do that which excepts a list of already printed attribues. Anyway, just an FYI. Its not needed here.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Just some minor comments

File: iree/compiler/Dialect/VM/IR/VMOps.cpp (hunk 4 above)
Comment (ScottTodd) [reply iree/compiler/Dialect/VM/IR/VMOps.cpp:764]: This custom printer/parser already skips some of the default formatting with its attributes. This special handling for symbol visibility as a bare word op prefix is forked from upstream:
* [visibility parsing](https://github.com/llvm/llvm-project/blob/16b284e99e4bca8d1e571310f545f10512dc979b/mlir/lib/IR/FunctionImplementation.cpp#L197-L198)
* [visibility printing](https://github.com/llvm/llvm-project/blob/16b284e99e4bca8d1e571310f545f10512dc979b/mlir/lib/IR/FunctionImplementation.cpp#L355-L357)

**Review by MaheshRavishankar — APPROVED** [review]

File: iree/compiler/Dialect/HAL/Target/TargetBackend.cpp (hunk 2 above)
Comment (ScottTodd) [reply iree/compiler/Dialect/HAL/Target/TargetBackend.cpp]: Took it a step further with
```c++
if (auto targetOp = targetSymbolMap[symbolName]) {
```

File: iree/compiler/Dialect/HAL/Target/TargetBackend.cpp (hunk 1 above)
Comment (ScottTodd) [reply iree/compiler/Dialect/HAL/Target/TargetBackend.cpp]: Ok, I expanded on the comment.

**Review by benvanik — APPROVED** [review]

**Pull request merged**